lotman/lotman_linux.go

```diff
@@ -314,9 +333,580 @@ func getFederationIssuer() (string, error) {
 	return federationIssuer, nil
 }
 
+// Given MPA1 and MPA2, merge them into a single MPA. If a field is set in MPA1,
+// it will take precedence.
+func mergeMPAs(mpa1, mpa2 *MPA) *MPA {
+	// Handle nil cases
+    if mpa1 == nil && mpa2 == nil {
+        return nil
+    }
+    if mpa1 == nil {
+        return mpa2
+    }
+    if mpa2 == nil {
+        return mpa1
+    }
+
+	// Merge individual fields
+    mergedMPA := *mpa1
+    if mpa1.DedicatedGB == nil {
+        mergedMPA.DedicatedGB = mpa2.DedicatedGB
+    }
+    if mpa1.OpportunisticGB == nil {
+        mergedMPA.OpportunisticGB = mpa2.OpportunisticGB
+    }
+    if mpa1.MaxNumObjects == nil {
+        mergedMPA.MaxNumObjects = mpa2.MaxNumObjects
+    }
+    if mpa1.CreationTime == nil {
+        mergedMPA.CreationTime = mpa2.CreationTime
+    }
+    if mpa1.ExpirationTime == nil {
+        mergedMPA.ExpirationTime = mpa2.ExpirationTime
+    }
+    if mpa1.DeletionTime == nil {
+        mergedMPA.DeletionTime = mpa2.DeletionTime
+    }
+
+    return &mergedMPA
+}
+
+// Given lot1 and lot2, merge them into a single lot. If a field is set in lot1,
+// it will take precedence. Lots cannot be merged if they have separate names
+func mergeLots(lot1, lot2 Lot) (Lot, error) {
+	if lot1.LotName != lot2.LotName {
+		return Lot{}, errors.Errorf("cannot merge lots %s and %s because they have different names", lot1.LotName, lot2.LotName)
+	}
+	mergedLot := lot1
+
+	// Prefer lot1's owner
+    if lot1.Owner == "" {
+        mergedLot.Owner = lot2.Owner
+    }
+
+	// Calculate union between the parents -- if this gets us in trouble by introducing cycles,
+	// lotman will tell us on startup (hopefully...).
+	parentSet := make(map[string]bool)
+    for _, parent := range lot1.Parents {
+        parentSet[parent] = true
+    }
+    for _, parent := range lot2.Parents {
+        if !parentSet[parent] {
+            mergedLot.Parents = append(mergedLot.Parents, parent)
+            parentSet[parent] = true
+        }
+    }
+
+	// Merge the MPAs
+	mergedLot.MPA = mergeMPAs(lot1.MPA, lot2.MPA)
+
+    return mergedLot, nil
+}
+
+// Calculate the union of two lot maps. If a lot is present in both maps, merge them.
+func mergeLotMaps(map1, map2 map[string]Lot) (map[string]Lot, error) {
+    result := make(map[string]Lot)
+
+    // Add all entries from map1 to result
+    for key, value := range map1 {
+        result[key] = value
+    }
+
+    // Merge entries from map2 into result
+    for key, value := range map2 {
+        if existingValue, exists := result[key]; exists {
+			mergedLot, err := mergeLots(existingValue, value)
+			if err != nil {
+				return result, err
+			}
+            result[key] = mergedLot
+        } else {
+            result[key] = value
+        }
+    }
+
+    return result, nil
+}
+
+// Grab a map of policy definitions from the config file, where the policy
+// name is the key and its attributes comprise the value.
+func getPolicyMap() (map[string]PurgePolicy, error) {
```

Comment:
Is there any sort of validation check that the values set in Lotman.PolicyDefinitions make sense, conform to the structure defined in parameters.yaml? That just says it's expecting and Object, so couldn't a use put anything there an it would propagate?

It would probably break things later, but would it give a meaningful/understandable error?

Reply:
No, you're right. I thought of this too and punted because of time constraints. That'll only hurt me in the long run, so I'll put together some validation code.